02-express-tutorial/app.js

```diff
@@ -1 +1,55 @@
-console.log('Express Tutorial')
+console.log("Express Tutorial");
+
+const express = require("express");
+const app = express();
+const { products } = require("./data.js");
+
+app.use(express.static("./public"));
+
+app.get("/api/v1/test", (req, res) => {
+  res.json({ message: "It worked!" });
+});
+// When a client sends a GET request to this URL,
+// the function (req, res) => { res.json(products); } is called. This function sends a JSON response to the client with the products array.
+app.get("/api/v1/products", (req, res) => {
+  res.json(products);
+});
+
+app.get("/api/v1/products/:productID", (req, res) => {
+  const idToFind = parseInt(req.params.productID);
+  const product = products.find((p) => p.id === idToFind);
+
+  if (product === undefined) {
+    res.status(404).json({ message: "That product was not found." });
+  } else {
+    res.json(product);
+  }
+});
+
+
+app.get('/api/v1/query', (req, res) => {
+    let results = products;
+  
+    if (req.query.search) {
+      const searchTerm = req.query.search.toLowerCase();
+      results = results.filter((p) => p.name.toLowerCase().startsWith(searchTerm));
+    }
+  
+    if (req.query.limit) {
+      const limit = parseInt(req.query.limit);
+      results = results.slice(0, limit);
+    }
+    if (req.query.maxPrice) {
+        const maxPrice = parseFloat(req.query.maxPrice);
+        results = results.filter((p) => p.price <= maxPrice);
+      }
+  
```

Comment:
Requires change:
This logic can be simplified using [Array.reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).
```js
app.get('/api/v1/query', (req, res) => {
    const { search, limit = 0, maxPrice = 0 } = req.query;

    // use Array.reduce to build a list of filtered products
    const filteredProducts = products.reduce((acc, product) => {
        // if the product price is greater than maxPrice OR
        // if there's a search and the product name doesn't include the search term
        // then return the accumulator (acc === list of filtered products)
        if (
            product.price > parseFloat(maxPrice) || 
            (search && !product.name.includes(search))
        ) {
            return acc;
        }

        // if the accumulator (acc === list of filtered products) is less than the limit, add product to the list
        if (limit && acc.length < parseInt(limit, 10)) {
            acc.push(product);
        }

        // return the accumulator (acc === list of filtered products) to check the next product
        return acc;
    }, []);

    res.status(200).json(filteredProducts);
});
```